relayer/src/services/polygon-zkevm.ts

```diff
@@ -0,0 +1,322 @@
+  1101: {
+    l1: {
+      PolygonZKEvmBridge: "0x2a3DD3EB832aF982ec71669E178424b10Dca2EDe",
+    },
+    l2: {
+      PolygonZKEvmBridge: "0x2a3DD3EB832aF982ec71669E178424b10Dca2EDe",
+    },
+    bridgeServiceUrl: "https://bridge-api.zkevm-rpc.com",
+  },
+  1442: {
+    l1: {
+      PolygonZKEvmBridge: "0xF6BEEeBB578e214CA9E23B0e9683454Ff88Ed2A7",
+    },
+    l2: {
+      PolygonZKEvmBridge: "0xF6BEEeBB578e214CA9E23B0e9683454Ff88Ed2A7",
+    },
+    bridgeServiceUrl: "https://bridge-api.public.zkevm-test.net",
+  },
+};
+export class PolygonZKEvmRelayerService extends BaseRelayerService {
+  MERKLE_PROOF_ENDPOINT = "/merkle-proof";
+  BRIDGES_ENDPOINT = "/bridge";
+  customNetworkContracts: NetworkContracts;
+
+  async getNetworkConfig(): Promise<NetworkContracts> {
+    const l2NetworkId = await this.getL2ChainId();
+    return this.customNetworkContracts ?? CONTRACTS[l2NetworkId];
+  }
+
+  async getMessageMerkleProof(
+    depositCount: number,
+    originNetwork: number,
+  ): Promise<MerkleProofResponse | undefined> {
+    const networkConfig = await this.getNetworkConfig();
+
+    const merkleProofUrl = new URL(networkConfig.bridgeServiceUrl + this.MERKLE_PROOF_ENDPOINT);
+    merkleProofUrl.searchParams.set("deposit_cnt", depositCount.toString());
+    merkleProofUrl.searchParams.set("net_id", originNetwork.toString());
+
+    const response = await fetch(merkleProofUrl);
+    const responseJson = await response.json();
+
+    if (responseJson.code && responseJson.message) {
+      throw new Error(responseJson.message);
+    }
+
+    return responseJson;
+  }
+
+  async getMessageInfoSafe(
+    depositCount: number,
+    originNetwork: number /** 1 means Polygon ZkEVM (L2) and 0 means L1 */,
+  ): Promise<PolygonZKEvmDeposit> {
+    // Loop and retry until the endpoint returns a proper result;
+    try {
+      const response = await this.getMessageInfo(depositCount, originNetwork);
+      return response;
+    } catch (e) {
+      if (e.message == "not found in the Storage") {
+        await sleep(2000);
+        return this.getMessageInfoSafe(depositCount, originNetwork);
+      }
+      throw e;
+    }
+  }
+
+  async getMessageInfo(
+    depositCount: number,
+    originNetwork: number /** 1 means Polygon ZkEVM (L2) and 0 means L1 */,
+  ): Promise<PolygonZKEvmDeposit> {
+    const networkConfig = await this.getNetworkConfig();
+
+    const messageBridgeUrl = new URL(networkConfig.bridgeServiceUrl + this.BRIDGES_ENDPOINT);
+    messageBridgeUrl.searchParams.set("deposit_cnt", depositCount.toString());
+    messageBridgeUrl.searchParams.set("net_id", originNetwork.toString());
+
+    const response = await fetch(messageBridgeUrl);
+    const responseJson = await response.json();
+
+    if (responseJson.code && responseJson.message) {
+      throw new Error(responseJson.message);
+    }
+
+    return responseJson.deposit;
+  }
+
+  async prepare(): Promise<boolean> {
+    return true;
+  }
```

Comment:
This is basically the `prepare` step of the relay service. You will learn more about it when the docs arrive.

In short, a relayer program for a message relay from rollup A -> rollup B, does the following:
- Constructs the relayers RelayerA & RelayerB
- Calls `RelayerB.prepare`
- Calls `RelayerA.relayTxToL1`
- Calls `RelayerB.finalize`

In depth:
1. RelayerB has to take care of preparing the right conditions so the message can be executed on L1 and accepted by the corresponding "Inbox" component that is supposed to move the message from L1 -> L2(rollup B)
2. RelayerA, after the RelayerB prepared the right conditions, executes the message coming from rollup A on L1
3. RelayerB, after the RelayerA executed the message on L1, makes sure that the message arrived on rollup B.

So, sometimes you will see that the `prepare` & `finalize` step are empty for some of the relayer services because there is no action required for these steps.
In the case of Optimism and Boba, their relayer component takes care of claiming/finalizing the message traveling to their rollup.

Reply:
thanks!